src/sunadaptcontroller/imexgus/sunadaptcontroller_imexgus.c

```diff
@@ -0,0 +1,222 @@
+{ return SUN_ADAPTCONTROLLER_H; }
+
+int SUNAdaptController_EstimateStep_ImExGus(SUNAdaptController C, sunrealtype h,
+                                            int p, sunrealtype dsm, sunrealtype* hnew)
+{
+  /* order parameter to use */
+  const int ord = p + 1;
+
+  /* modified method for first step */
+  if (SACIMEXGUS_FIRSTSTEP(C))
+  {
+    /* set usable time-step adaptivity parameters */
+    const sunrealtype k = -RCONST(1.0) / ord;
+    const sunrealtype e = SUNMAX(SACIMEXGUS_BIAS(C) * dsm, TINY);
+
+    /* compute estimated optimal time step size */
+    *hnew = h * SUNRpowerR(e,k);
+  }
+  else
+  {
+    /* set usable time-step adaptivity parameters */
+    const sunrealtype k1e = -SACIMEXGUS_K1E(C) / ord;
+    const sunrealtype k2e = -SACIMEXGUS_K2E(C) / ord;
+    const sunrealtype k1i = -SACIMEXGUS_K1I(C) / ord;
+    const sunrealtype k2i = -SACIMEXGUS_K2I(C) / ord;
+    const sunrealtype e1 = SUNMAX(SACIMEXGUS_BIAS(C) * dsm, TINY);
+    const sunrealtype e2 = e1 / SUNMAX(SACIMEXGUS_EP(C), TINY);
+    const sunrealtype hrat = h / SACIMEXGUS_HP(C);
+
+    /* compute estimated optimal time step size */
+    *hnew = h * SUNMIN(hrat * SUNRpowerR(e1,k1i) * SUNRpowerR(e2,k2i),
+                       SUNRpowerR(e1,k1e) * SUNRpowerR(e2,k2e));
+  }
+
+  /* return with success */
+  return SUNADAPTCONTROLLER_SUCCESS;
+}
+
+int SUNAdaptController_Reset_ImExGus(SUNAdaptController C)
+{
+  SACIMEXGUS_EP(C) = RCONST(1.0);
+  SACIMEXGUS_FIRSTSTEP(C) = SUNTRUE;
+  return SUNADAPTCONTROLLER_SUCCESS;
+}
+
+int SUNAdaptController_SetDefaults_ImExGus(SUNAdaptController C)
+{
+  SACIMEXGUS_K1E(C)    = DEFAULT_K1E;
+  SACIMEXGUS_K2E(C)    = DEFAULT_K2E;
+  SACIMEXGUS_K1I(C)    = DEFAULT_K1I;
+  SACIMEXGUS_K2I(C)    = DEFAULT_K2I;
+  SACIMEXGUS_BIAS(C)   = DEFAULT_BIAS;
+  return SUNADAPTCONTROLLER_SUCCESS;
+}
+
+int SUNAdaptController_Write_ImExGus(SUNAdaptController C, FILE *fptr)
+{
+  fprintf(fptr, "ImEx Gustafsson SUNAdaptController module:\n");
+#if defined(SUNDIALS_EXTENDED_PRECISION)
+  fprintf(fptr, "  k1e = %32Lg\n", SACIMEXGUS_K1E(C));
+  fprintf(fptr, "  k2e = %32Lg\n", SACIMEXGUS_K2E(C));
+  fprintf(fptr, "  k1i = %32Lg\n", SACIMEXGUS_K1I(C));
+  fprintf(fptr, "  k2i = %32Lg\n", SACIMEXGUS_K2I(C));
+  fprintf(fptr, "  bias factor = %22Lg\n", SACIMEXGUS_BIAS(C));
+  fprintf(fptr, "  previous error = %22Lg\n", SACIMEXGUS_EP(C));
+  fprintf(fptr, "  previous step = %22Lg\n", SACIMEXGUS_HP(C));
+#else
+  fprintf(fptr, "  k1e = %16g\n", SACIMEXGUS_K1E(C));
+  fprintf(fptr, "  k2e = %16g\n", SACIMEXGUS_K2E(C));
+  fprintf(fptr, "  k1i = %16g\n", SACIMEXGUS_K1I(C));
+  fprintf(fptr, "  k2i = %16g\n", SACIMEXGUS_K2I(C));
+  fprintf(fptr, "  bias factor = %16g\n", SACIMEXGUS_BIAS(C));
+  fprintf(fptr, "  previous error = %16g\n", SACIMEXGUS_EP(C));
+  fprintf(fptr, "  previous step = %16g\n", SACIMEXGUS_HP(C));
+#endif
+  return SUNADAPTCONTROLLER_SUCCESS;
+}
+
+int SUNAdaptController_SetErrorBias_ImExGus(SUNAdaptController C, sunrealtype bias)
+{
```

Comment:
```suggestion
{
  if (C == NULL) { return SUNADAPTCONTROLLER_ILL_INPUT; }
```

Reply:
Although it's not the typical use case, this function can be called directly so I think we should still check the inputs.